aptos-move/move-examples/shared_account/sources/SharedAccount.move

```diff
@@ -0,0 +1,195 @@
+            denominator2
+        }
+        else {
+            find_greatest_common_divisor(denominator2 % denominator1, denominator1)
+        }
+    }
+
+    fun find_least_common_denominator(denominator1: u64, denominator2: u64) : u64 {
+        assert!(denominator1 * denominator2 > 0, Errors::invalid_argument(EINVALID_INPUT));
+        let gcd: u64 = find_greatest_common_divisor(denominator1, denominator2);
+        assert!(gcd > 0, Errors::invalid_argument(Errors::invalid_argument(EINVALID_INPUT)));
+        (denominator1 * denominator2) / gcd
+    }
+
+    // Create and initialize a shared account
+    public fun initialize(source: &signer, seed: vector<u8>, addresses: vector<address>, numerators: vector<u64>, denominators: vector<u64>): address acquires SharedAccount {
+        assert!(Vector::length(&addresses) == Vector::length(&numerators), Errors::invalid_argument(EINVALID_INPUT));
+        assert!(Vector::length(&addresses) == Vector::length(&denominators), Errors::invalid_argument(EINVALID_INPUT));
+
+        assert!(Account::exists_at(Signer::address_of(source)), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+        let (resource_signer, resource_signer_cap) = Account::create_resource_account(source, seed);
+        if (!exists<SharedAccount>(Signer::address_of(&resource_signer))) {
+            move_to(
+                &resource_signer,
+                SharedAccount {
+                    share_record: Vector::empty<State>(),
+                    signer_capability: resource_signer_cap,
+                }
+            )
+        };
+
+        let shared_account = borrow_global_mut<SharedAccount>(Signer::address_of(&resource_signer));
+
+        let i = 0;
+        let cumulative_numerator = 0;
+        let cumulative_denominator = 1;
+
+        while (i < Vector::length(&addresses)) {
+            let num = *Vector::borrow(&numerators, i);
+            let denom = *Vector::borrow(&denominators, i);
+            let addr = *Vector::borrow(&addresses, i);
+            assert!(Account::exists_at(addr), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+            assert!(num >= 0 && denom > 0, Errors::invalid_argument(EINVALID_INPUT));
+
+            let common_denom = find_least_common_denominator(cumulative_denominator, denom);
+            cumulative_numerator = cumulative_numerator * (common_denom / cumulative_denominator) + num * (common_denom / denom);
+            cumulative_denominator = common_denom;
+
+            Vector::push_back(&mut shared_account.share_record, State { share_holder: addr, numerator: num, denominator: denom });
+            i = i + 1;
+        };
+
+        assert!(cumulative_numerator == cumulative_denominator, Errors::invalid_argument(EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS));
+        Signer::address_of(&resource_signer)
+    }
+
+    // Disperse all available balance to addresses in the shared account
+    public(script) fun disperse<CoinType>(resource_addr: address) acquires SharedAccount {
```

Comment:
suppose A has 2/3 and B has 1/3, A call this function every time there are two coins. B would never has any coin. I think we may want to limit the call when the there is a large enough balance, 10 * num_holder_accounts

Reply:
That's a reasonable idea, but a few thoughts:
* This is more of a "here's a method for doing this"
* our decimal is either 8 or 10, so anyone splitting 2 coins is well.. spending more on gas
* the initial layout will dictate the distribution, so someone could just as well do 1/3 and 2/3 to improve distribution of the right most value.

Reply:
The more general problem is the person with the biggest share can deprive all people with smaller shares all their fund. The return of this attack is = number_of_people_in_the_pool - 1. suppose we have 100 people in the pool, the return is 99 coins and the cost is the gas fee of the transaction. If the value is bigger than the cost, they can keep abusing the this function call. 


How about adding a one line assert assert!(balance >= sum_of_share, EBALANCE_SMALLER_OR_EQUAL_TO_TOTOAL_SHARES)? If A has 2 shares and B has 1 share, the sum_of_share is 3 this case.